// thirdweb/dist/esm/extensions/prebuilts/get-required-transactions.js
import { getDeployedCreate2Factory } from "../../contract/deployment/utils/create-2-factory.js";
import { getDeployedInfraContract } from "../../contract/deployment/utils/infra.js";
import { getDeployedInfraContractFromMetadata } from "../../contract/deployment/utils/infra.js";
import { computePublishedContractAddress } from "../../utils/any-evm/compute-published-contract-address.js";
/**
 * @internal
 */
export async function getRequiredTransactionCount(options) {
    const { chain, client, metadata, implementationConstructorParams, modules = [], } = options;
    if (metadata?.deployType === "autoFactory") {
        const results = await Promise.all([
            getDeployedCreate2Factory({
                chain,
                client,
            }).then((c) => c ? null : { type: "infra", contractId: "Create2Factory" }),
            getDeployedInfraContract({
                chain,
                client,
                contractId: "Forwarder",
            }).then((c) => c ? null : { type: "infra", contractId: "Forwarder" }),
            getDeployedInfraContract({
                chain,
                client,
                contractId: "TWCloneFactory",
                constructorParams: {
                    _trustedForwarder: await computePublishedContractAddress({
                        chain,
                        client,
                        contractId: "Forwarder",
                    }),
                },
            }).then((c) => c ? null : { type: "infra", contractId: "TWCloneFactory" }),
            // TODO (deploy): add WETH contract check for implementations that need it (check implementation constructor params)
            getDeployedInfraContract({
                chain,
                client,
                contractId: metadata.name,
                constructorParams: implementationConstructorParams,
                publisher: metadata.publisher,
                version: metadata.version,
            }).then((c) => c
                ? null
                : {
                    type: "implementation",
                    contractId: metadata.name,
                }),
            ...modules.map((m) => getDeployedInfraContractFromMetadata({
                chain,
                client,
                contractMetadata: m,
            }).then((c) => c
                ? null
                : {
                    type: "module",
                    contractId: m.name,
                })),
        ]);
        results.push({ type: "proxy", contractId: metadata.name });
        return results.filter((r) => r !== null);
    }
    return [{ type: "implementation", contractId: metadata.name }];
}
